File: backend/src/main/java/observer/quantum/worm/domain/project/element/ProjectElementDto.java
package observer.quantum.worm.domain.project.element;

import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Schema(description = "Project element data transfer object")
public class ProjectElementDto {
  @Schema(description = "Unique identifier of the element")
  private String id;

  @Schema(description = "Version for optimistic locking")
  private Long version;

  @NotBlank(message = "Name is required")
  @Schema(description = "Name of the element", required = true)
  private String name;

  @NotNull(message = "Type is required")
  @Schema(description = "Type of the element (FOLDER/ITEM)", required = true)
  private ElementType type;

  @Schema(description = "Position for ordering elements")
  @NotNull(message = "Position is required")
  private Integer position;

  @Schema(description = "Level in the tree hierarchy")
  @NotNull(message = "Level is required")
  private Integer level;

  // Client-side computed properties
  @Schema(description = "Whether the element can be expanded (computed from type)")
  private Boolean expandable;

  public ProjectElementDto(ProjectElement element) {
    this.id = element.getId();
    this.version = element.getVersion();
    this.name = element.getName();
    this.type = element.getType();
    this.position = element.getPosition();
    this.level = element.getLevel();
    this.expandable = element.getType().isExpandable();
  }

  public ProjectElement toProjectElement() {
    return ProjectElement.builder()
        .id(id)
        .version(version)
        .name(name)
        .type(type)
        .position(position)
        .level(level)
        .build();
  }
}

File: worm-server/src/project/element/project-element.dto.ts
// project-element.dto.ts
import { ApiProperty } from '@nestjs/swagger';
import { IsNotEmpty, IsString, IsEnum, IsNumber } from 'class-validator';
import { ElementType, isExpandable } from './element-type.enum.js';

export class ProjectElementDto {
  @ApiProperty({
    description: 'Unique identifier of the element',
    example: 'd42200be-2a40-4c3e-9c35-47c8f641c8ea',
    required: false,
  })
  id?: string;

  @ApiProperty({
    description: 'Version for optimistic locking',
    required: false,
  })
  version?: number;

  @ApiProperty({ description: 'Name of the element', example: 'Chapter 1' })
  @IsNotEmpty({ message: 'Name is required' })
  @IsString()
  name: string;

  @ApiProperty({
    description: 'Type of the element (FOLDER/ITEM)',
    enum: ElementType,
    example: ElementType.FOLDER,
  })
  @IsEnum(ElementType, {
    message: 'Type is required and must be FOLDER or ITEM',
  })
  type: ElementType;

  @ApiProperty({
    description: 'Position for ordering elements',
    example: 1,
  })
  @IsNotEmpty({ message: 'Position is required' })
  @IsNumber()
  position: number;

  @ApiProperty({
    description: 'Level in the tree hierarchy',
    example: 0,
  })
  @IsNotEmpty({ message: 'Level is required' })
  @IsNumber()
  level: number;

  @ApiProperty({
    description: 'Whether the element can be expanded (computed from type)',
    required: false,
  })
  expandable?: boolean;

  constructor(entity?: ProjectElementDto) {
    if (entity) {
      this.id = entity.id;
      this.version = entity.version;
      this.name = entity.name;
      this.type = entity.type;
      this.position = entity.position;
      this.level = entity.level;
      this.expandable = isExpandable(entity.type);
    }
  }
}
